### PR TITLE
Update legacy context api

### PR DESCRIPTION
**Pull request by PiotrKozlowski**

### CODE REVIEW

File: src/slate/context.js [hunk 1]
```diff
@@ -68,14 +60,24 @@ export default class CounterContext extends React.Component {
...
     }
 }
 
-CounterContext.childContextTypes = {
-    counters: PropTypes.instanceOf(Immutable.Map),
-    counterUpdater: PropTypes.instanceOf(CounterContext),
-}
+export const CounterContext = React.createContext({
+    counters: new Immutable.Map(),
+    counterUpdater: new CountersProvider(),
```
Comment (aiwenar) [inline src/slate/context.js]: I don't think manually creating React components is supported, and I'm not comfortable with it.

**Review by aiwenar — CHANGES_REQUESTED** [review]
If we're migrating to the new context API then there should be no need for `CounterContext` / `CountersProvider` any more. It was only used to ensure that consumers are updated when counters change, which could not happen due to a `componentShouldUpdate` somewhere between the provider and the consumer. In fact, this pattern has been [cited](https://github.com/reactjs/rfcs/blob/master/text/0002-new-version-of-context.md#shifts-complexity-to-user-space) as one of reasons for introducing the new API.

If I understand correctly how the new API works, all this could be simplified to just

```js
const Counters = React.createContext(new Immutable.Map())
export default Counters

export const WithCounters = getKey => Component => {
    function WithCounters(props) {
        const key = getKey(props)
        return <Counters.Consumer>
            {counters => <Component
                counters={counters.get(key, new Immutable.Map())}
                {...props}
                />
            }
        </Counters.Consumer>
    }

    WithCounters.displayName = `WithCounters(${
        Component.displayName || Component.name
    })`

    return WithCounters
}

File: src/slate/context.js [hunk 2]
```diff
@@ -2,80 +2,11 @@
...
-        return this.props.children
-    }
-}
-
-CounterContext.childContextTypes = {
-    counters: PropTypes.instanceOf(Immutable.Map),
-    counterUpdater: PropTypes.instanceOf(CounterContext),
-}
+const CountersContext = React.createContext(new Immutable.Map())
+export default CountersContext
```
Comment (aiwenar) [inline src/slate/context.js]: I see no reason for this name change. If you do want to change the name then I suggest `Counters`; `Counters.Provider` and `Counters.Consumer` seem much nicer than `CounterContext.Provider` and `CounterContext.Consumer`.

Comment (PiotrKozlowski) [reply src/slate/context.js]: But `Counters` doesn't provide any information about what is it. `NameContext` is standard for naming context

**Comment by aiwenar** [comment]
Please base this PR on master, not use-onsetvalue.

**Comment by PiotrKozlowski** [comment]
Done

**Pull request merged**